acme/challenge.go

```diff
@@ -342,6 +354,258 @@ func dns01Validate(ctx context.Context, ch *Challenge, db DB, jwk *jose.JSONWebK
+			"error retrieving claims from ID token"))
+	}
+
+	challengeValues, err := wire.ParseID([]byte(ch.Value))
+	if err != nil {
+		return WrapErrorISE(err, "error unmarshalling challenge data")
+	}
+
+	expectedKeyAuth, err := KeyAuthorization(ch.Token, jwk)
+	if err != nil {
+		return err
+	}
+	if expectedKeyAuth != wireChallengePayload.KeyAuth {
+		return storeError(ctx, db, ch, true, NewError(ErrorRejectedIdentifierType,
+			"keyAuthorization does not match; expected %s, but got %s", expectedKeyAuth, wireChallengePayload.KeyAuth))
+	}
+
+	if challengeValues.Name != claims.Name || challengeValues.Handle != claims.Handle {
+		return storeError(ctx, db, ch, false, NewError(ErrorRejectedIdentifierType, "OIDC claims don't match"))
+	}
+
+	// Update and store the challenge.
+	ch.Status = StatusValid
+	ch.Error = nil
+	ch.ValidatedAt = clock.Now().Format(time.RFC3339)
+
+	if err = db.UpdateChallenge(ctx, ch); err != nil {
+		return WrapErrorISE(err, "error updating challenge")
+	}
+
+	parsedIDToken, err := jwt.ParseSigned(wireChallengePayload.IDToken)
+	if err != nil {
+		return WrapErrorISE(err, "Invalid OIDC id token")
+	}
+	oidcToken := make(map[string]interface{})
+	if err := parsedIDToken.UnsafeClaimsWithoutVerification(&oidcToken); err != nil {
+		return WrapErrorISE(err, "Failed parsing OIDC id token")
+	}
+
+	orders, err := db.GetAllOrdersByAccountID(ctx, ch.AccountID)
+	if err != nil {
+		return WrapErrorISE(err, "Could not find current order by account id")
+	}
+
+	if len(orders) == 0 {
+		return WrapErrorISE(err, "There are not enough orders for this account for this custom OIDC challenge")
+	}
+
+	order := orders[len(orders)-1]
+
+	if err := db.CreateOidcToken(ctx, order, oidcToken); err != nil {
+		return WrapErrorISE(err, "Failed storing OIDC id token")
+	}
+
+	return nil
+}
+
+func wireDPOP01Validate(ctx context.Context, ch *Challenge, db DB, jwk *jose.JSONWebKey, payload []byte) error {
+	prov, ok := ProvisionerFromContext(ctx)
+	if !ok {
+		return NewErrorISE("missing provisioner")
+	}
+
+	rawKid, thumbprintErr := jwk.Thumbprint(crypto.SHA256)
+	if thumbprintErr != nil {
+		return storeError(ctx, db, ch, false, WrapError(ErrorServerInternalType, thumbprintErr, "failed to compute JWK thumbprint"))
+	}
+
+	kid := base64.RawURLEncoding.EncodeToString(rawKid)
+
+	dpopOptions := prov.GetOptions().GetDPOPOptions()
+	key := dpopOptions.GetSigningKey()
+
+	var wireChallengePayload WireChallengePayload
+	err := json.Unmarshal(payload, &wireChallengePayload)
+	if err != nil {
+		return storeError(ctx, db, ch, false, WrapError(ErrorRejectedIdentifierType, err,
+			"error unmarshalling Wire challenge payload"))
+	}
+
+	file, err := os.CreateTemp(os.TempDir(), "acme-validate-challenge-pubkey-")
+	if err != nil {
+		return WrapErrorISE(err, "temporary file could not be created")
+	}
+	defer file.Close()
+	defer os.Remove(file.Name())
+
+	buf := bytes.NewBuffer(nil)
+	buf.WriteString(key)
+
+	n, err := file.Write(buf.Bytes())
```

Comment:
Seems unnecessary:
```suggestion
	buf := bytes.NewBuffer(nil)
	buf.WriteString(key)

	n, err := file.Write([]byte(key))
```

Btw this also exists `bytes.NewBufferString("foo")`

Reply:
This will be gone with https://github.com/smallstep/certificates/pull/1670